### DISCUSSION TITLE
is there a way to use a soundfont not provided by magenta.js?

### DISCUSSION THREAD

**Discussion by nullcasting** [cifkao/html-midi-player#43] [Q&A] [answered]
Hello! I'd like to embed a roland-esque soundfont onto my site to use with this lovely tool, but i can't seem to get ti to work. Is there functionality to do this?

**Comment by cifkao** [comment]
Do you mean using a custom soundfont with the player? That would require converting it to the right format. I believe @majorcyto has done this for [musicalnexus.net](https://musicalnexus.net/).

**Reply by nullcasting** [reply]
yes, that's what i mean! what's the correct format? the file i was trying to use is .sf2 which im assuming is incorrect lol

**Reply by cifkao** [reply]
The format seems to be designed by Magenta and it is described here:
https://github.com/magenta/magenta-js/blob/d8a76682abb0979b985e4b80f6b68b5123b9f8d5/music/src/core/soundfont.ts#L381-L419

Not sure if there is a conversion script from SF2; probably not since this format is much more limited than SF2.

**Reply by nullcasting** [reply]
ah, i'll take a look at this. thank you

**Reply by majorcyto** [reply]
Hi,

That is my website, and yes it uses a specific format as explained by Magenta Docs. I had to separate out all of the notes into individual files and group them by instrument. Kind of painful depending on the tooling you have available tbh. Mine was half manual, half scripted due to being limited in how I could extract the instrument sounds I used from the game. I also had to create variations in my specific use case so I ended up creating around 58,000+ note files lol.

Hopefully you don't need so many instruments / ranges!

**Comment by valentijnnieman** [comment, marked as answer]
Hey! Was redirected here from #62 - what I did was modify the script found here: https://github.com/mudcube/MIDI.js/blob/master/generator/ruby/soundfont_builder.rb which turns a `sf2` file into a format used for MIDI.js, to the format Magenta expects. It seems to work for me using this library, so thought I'd share: https://gist.github.com/valentijnnieman/00d93627a0da7d68a836553748f73564 note the comment that says to also manually create a `soundfont.json` file!

**Reply by cifkao** [reply]
Wow, that's awesome! Thanks for sharing this!

**Reply by valentijnnieman** [reply]
No problem! I should note though this is a very quick and dirty solution. It doesn't properly set the `minPitch` and `maxPitch` variables for the individual `instrument.json` files, for example, which I just noticed could throw an error if trying to play a pitch that doesn't have a `.mp3` file in the instrument folder (probably because the sf2 didn't support that pitch?).

**Reply by pjr22** [reply]
Many thanks @valentijnnieman! With minimal effort (mostly installing ruby and necessary dependencies) I was able to convert a sizable sf2 soundfont to the necessary file structure to work with magenta/html-midi-player.  On the first pass I did run into the "Broken table" error on table validation, and not knowing what to do about it, I just commented out the validation check and crossed my fingers.  Fortunately it worked!  You're a rock star!

**Reply by valentijnnieman** [reply]
Thanks @pjr22 ! Happy it was helpful!

**Reply by Guseyn** [reply]
I expanded the script for generating notes with different velocities. From the source code in magneta.js I could see that it's enough to generate notes with just most common velocities, since the script always tries to find closest velocity which is specified in the `instruments.json` files.

Here is the gist: https://gist.github.com/Guseyn/2946a3dd502f7168ca8d9c7e24b67f81
